### PR TITLE
Restrict auto-slug generation to first non-empty block

### PR DESCRIPTION
**Pull request by bfollington**
Fixes https://github.com/subconsciousnetwork/subconscious/issues/817
Related to https://github.com/subconsciousnetwork/subconscious/issues/804

- Introduce Subtext.generateSlug
- Reduce excerpt length, it's too long in practice imo
- Introduce `truncatedPrefix()` to add ellipsis (…) if string is truncated

280 char excerpt:

image

512 char excerpt (old value):

image

### CODE REVIEW

File: xcode/Subconscious/Shared/Library/StringUtilities.swift [hunk 1]
```diff
@@ -74,3 +74,14 @@ extension StringProtocol {
...
+extension String {
+    /// Similar to `prefix` but inserts an ellipsis (…) if truncation occurs
+    func truncatedPrefix(_ count: Int) -> String {
+        let truncated =
+            String(self.prefix(count - (self.count > count ? 1 : 0)))
+                .trimmingCharacters(in: .whitespacesAndNewlines)
+        
+        return self.count > count ? truncated + "…" : truncated
+    }
+}
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Library/StringUtilities.swift:87]: We already have `String.truncate` in `Shared/Library/Prose.swift`. Does this not cover the same usecase?

File: xcode/Subconscious/Shared/Parsers/Subtext.swift [hunk 2]
```diff
@@ -663,10 +663,30 @@ extension Subtext {
         return output.isEmpty ? fallback : output
     }
     
+    static let maxExcerptSize = 280
+    static let maxGeneratedSlugSize = 128
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Parsers/Subtext.swift:667]: These should probably be private

**Review by gordonbrander — CHANGES_REQUESTED** [review]
Small change requested. We already have a truncation utility function.

File: xcode/Subconscious/Shared/Library/StringUtilities.swift (hunk 1 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Library/StringUtilities.swift]: ty! I had a feeling I'd done this before

**Review by gordonbrander — APPROVED** [review]
LGTM

**Pull request merged**